### ISSUE TITLE
US: Navigation gets off the freeway [San francisco]

### ISSUE THREAD

**Issue by dead10ck** [osmandapp/Osmand#9235]
This is with version 3.7.3, on OxygenOS. The map is the San Francisco map with OSM Live enabled.

There's a stretch of I880 South that gets off the freeway and then back on for seemingly no reason. To see it in action, route between 849VMV6V+XV and 849VJWX3+H6.

I've checked the map data, and I don't see anything wrong, like a disconnected way or anything like that. There are no traffic lights or any other things along the way that I could see causing this.

ORS does not do this.

![Screenshot_20200613-230800](https://user-images.githubusercontent.com/846275/84584315-bdf56400-add0-11ea-921a-030228ca8596.jpg)

**Comment by naoliv** [comment]
Probably because the highway https://www.openstreetmap.org/way/156115057 has `maxspeed` while the link doesn't https://www.openstreetmap.org/way/510431162

Then it's considered to be faster to use the link instead the highway (because the assumed default `maxspeed` is higher)

**Comment by dead10ck** [comment]
Hmm, what is the default maxspeed for a motorway link? In the US, they're usually very slow compared to the motorway.

But also using maxspeed in routing decisions seems like asking for trouble, as evident by this issue. Especially considering how spotty the presence of the maxspeed tag is in the wild. It should probably only be used in estimating time.

**Comment by scaidermern** [comment]
See a similar discussion at #8376.

**Comment by jfburke3** [comment]
@dead10ck This has been a longstanding problem. While the default maxspeed for a `motorway_link` is slightly lower than the default for a `motorway`, it's still higher than explicitly set maxspeed values in dense urban areas. A while back someone suggested making the OsmAnd default to be half of the default for motorway, but that never happened. (Personally I think that would be too low, anyway. Maybe 2/3 or 3/4 would be better.)

One thing you can do to help the problem is, if that exit ramp has a yellow speed sign, tag the _link with `maxspeed:advisory` and the value. OsmAnd will use that in place of a regulatory speed limit.

**Comment by vshcherb** [comment]
It looks like this issue is better to solve on map i.e. it is possible to put a restriction via route or to put a maxspeed limitation on motorway_link. 
Tweaking parameters for priorities motorway_link vs motorway has other negative effects.

**Comment by vshcherb** [comment]
maxspeed:practical could also be used

**Comment by dead10ck** [comment]
@vshcherb putting a maxspeed tag might solve this particular case, but this seems like a generalized problem. Practically speaking, we can't expect perfectly complete map data, so fixing this particular motorway junction would be great, but there will be others.

But in any case, even if it were true in the real world that a motorway link had a higher maxspeed, that does not mean this is the correct thing to do. No matter what, it should not be getting off the freeway and just to get back on, barring something very expensive, such as a car accident.

Since other routers do not do this, I see that as evidence that a solution must exist that does not take needless detours and accommodates things like the roundabout case.

**Comment by vshcherb** [comment]
Well it all needs investigation how many times it happens and so on. In general we do have maxspeed on links (same as on motorway) on https://www.openstreetmap.org/way/99243827. So it could be a generic problem for United States. 

Other routers could give extra penalty for motorway_link vs motorway but it simply could be not enough! Cause default motorway_link speed is  110 kmh / 68.35 mph and there is extra penalty. Anyway it all requires investigation how many times it happens and what actually causes it
```
<!-- most important backbone roads -->
			<select value="110" t="highway" v="motorway"/>
			<select value="110" t="highway" v="motorway_link"/>
<select value="1.1" t="highway" v="motorway"/>
			<!-- make links slightly smaller, so in large parallel intersections, ways will be preferred over links -->
			<select value="1.05" t="highway" v="motorway_link"/>

```

**Comment by dead10ck** [comment]
68 is way too high of a default for a motorway link, at least in the US. In my experience, most of the time they are usually 30–40mph, but sometimes even 20mph, depending on how short it is, or how sharp of a turn is required.

Even so though, regardless of speed, it should simply not choose to get off the freeway and back on. Cases where this is desirable is very rare.

**Comment by AllroadsNL** [comment]
**A other specific example:** with maxspeed difference on motorway and motorway_link

![Emmeloord D](https://user-images.githubusercontent.com/8817345/86615508-d2cead00-bfb4-11ea-9f53-f5156b728a3c.png)

Image situation: OSM website routing OSRM/Graphhopper

![Dtour motorway junction](https://user-images.githubusercontent.com/8817345/86605216-b0359780-bfa6-11ea-85aa-b1d5bc641914.JPG)

Espected:
That the motorway was taken straight on.

Failure:
The D tour is taken over the motorway_link.

Data is changed recently because of Government maxspeed rules on motorway. (nation wide)
motorway:
maxspeed=100 maxspeed:conditional=130 @ (19:00-06:00)

But on the motorway_link to the motorway  [situation streetview](https://www.google.nl/maps/@52.6828673,5.7426369,3a,75y,58.98h,85.5t/data=!3m6!1e1!3m4!1sW1wyZpWCd4n2l1WnIHN7cA!2e0!7i13312!8i6656) there is this  G01 trafficsign

![NLG01](https://user-images.githubusercontent.com/8817345/86606805-c3e1fd80-bfa8-11ea-94fe-7ace37b9c256.gif)

This means from that point,[ rules says you can drive 130](https://wiki.openstreetmap.org/wiki/NL:Overzicht_Nederlandse_Verkeersborden#G:_Verkeersregels). And Government did not adjust that,
Once directly on the motorway, we get mostly a trafficsign with maxspeed=100 maxspeed:conditional=130 @ (19:00-06:00), because there is a other junction comming up, in this case, there is only a maxspeed=100 sign.
[Stituation streetview](https://www.mapillary.com/map/im/LQJ7g-DB4tbbFM9dCjx3GA).

Maybe for this type of situations, with a motorway_link. where the maxspeed is higher as the motorway, where a D tour is visualised wrongly a other time/speed calculation should be used?

There is also highway=give_way on the exit motorway_link, that does not have the effect as it should have. To better calculate the time/speed ratio?

General:
At a give_way, people come mostly to a stop, so this effect the time/speed ratio.
At a entrance motorway_link, people have to accelarate mostly, this should effect the time/speed ratio. 

Or is the program struggling with the  conditional settings?

![Dtour motorway junctionJOSM](https://user-images.githubusercontent.com/8817345/86609177-dad61f00-bfab-11ea-9f78-6807b2cd992b.JPG)


The data in OSM is correct, a situation, that now is common in the Netherlands. 
The Government, says, because of the entrence of a motorway, people mostly drive/ride not that 130 maxspeed, must accelerate, because it is a entrance. More trafficsigns, does not make the road safer. After a cpuple of hunderd meters they get the new maxspeed sign.
It is a consideration of more trafficsigns, costs and safety risks.

And there appear more cases where it is not right.
![QDKV4Wt](https://user-images.githubusercontent.com/8817345/86615852-4ffa2200-bfb5-11ea-91ab-c928e69a272e.png)
![RpZu2wy](https://user-images.githubusercontent.com/8817345/86615860-538da900-bfb5-11ea-8d17-a7975a5e5069.png)

**Comment by vshcherb** [comment]
I can't reproduce Emmelord issue on latest map data, probably we need to wait July maps and try it again. give_away penalty should be enough here

**Comment by AllroadsNL** [comment]
> give_away penalty should be enough here

The give_way node is tagged there, 28-06-2018.
This did not solved the problem.

Recently a user had this failure. (screenshot)

**Comment by AllroadsNL** [comment]
> I can't reproduce Emmeloord issue on latest map data

It is late now, if you reproduce it in the evening after 19:00, then the maxspeed on the motorway is 130.
motorway:
maxspeed=100 maxspeed:conditional=130 @ (19:00-06:00)

Try to reproduce it, when, it is daytime.

Does that matter?

**Comment by vshcherb** [comment]
I put the time manually, well will double check.

**Comment by gunnar22** [comment]
Thanks for giving us OsmAnd. I use it a lot, it's great!

Examples from Norway. 

(OsmAnd+ 3.7.4, Android, latest maps).

To reproduce, create a DIRECTIONS route from Klett (63.32406, 10.30578) to Stjørdal (63.46709, 10.92340) (for driving by car). The problem occurs at five places. The route leaves the highway=trunk, goes to trunk_link, (to roundabout), to trunk_link, and then back to the same trunk, instead of staying on the trunk all the way.

- Heggstadmoen (63.33064, 10.35352).
- Vikhammerdalen (63.42231, 10.63092).
- Malvik storsenter (63.41924, 10.75631).
- Hommelvik (63.41261, 10.82056).
- Rotvoll (63.43062, 10.48791), something special happens here: We can see either correct or incorrect routing.
When starting from Klepp, the route is correct. But, when moving the starting point closer (e.g., 1 km away from this position) then the error occurs here too.

I think that it is important to fix this. One may 'blame' the map data, but anyway, unfortunately it makes the routing in OsmAnd appear 'stupid'. While driving, it can be annoying to listen to incorrect navigation voice prompts from OsmAnd asking me to exit the motorway. If I could set the priorities, I'd say that it would be worth some extra effort to solve this problem without being dependent on speed limit data on road links.

Here are some screen shots of the example described above:

![Route-Klett-Stjordal](https://user-images.githubusercontent.com/30504902/87246497-224a2880-c44e-11ea-8004-4bc1efd719c3.png)
Route from Klett to Stjørdal. Five places with incorrect routing, see detailed figures.

![Heggstadmoen](https://user-images.githubusercontent.com/30504902/87246591-b1efd700-c44e-11ea-98d5-aeaed9f11609.png)
Heggstadmoen (63.33064, 10.35352).

![Vikhammerdalen](https://user-images.githubusercontent.com/30504902/87246598-b7e5b800-c44e-11ea-98f1-a5159482281d.png)
Vikhammerdalen (63.42231, 10.63092).

![Malvik-storsenter](https://user-images.githubusercontent.com/30504902/87246601-bd430280-c44e-11ea-85ce-618530d535a3.png)
Malvik storsenter (63.41924, 10.75631).

![Hommelvik](https://user-images.githubusercontent.com/30504902/87246607-c0d68980-c44e-11ea-9cd1-f662431477ab.png)
Hommelvik (63.41261, 10.82056).

![Rotvoll-correct](https://user-images.githubusercontent.com/30504902/87246612-c633d400-c44e-11ea-8961-ce6fc6b84ccb.png)
Rotvoll (63.43062, 10.48791) with correct routing.

![Rotvoll](https://user-images.githubusercontent.com/30504902/87246613-c9c75b00-c44e-11ea-9610-5225ae1f1208.png)
Rotvoll (63.43062, 10.48791) with incorrect routing when the starting point is closer (not at Klett, but a few km south-west).

**Comment by dead10ck** [comment]
> I think that it is important to fix this. One may 'blame' the map data, but anyway, unfortunately it makes the routing in OsmAnd appear 'stupid'. While driving, it can be annoying to listen to incorrect navigation voice prompts from OsmAnd asking me to exit the motorway. If I could set the priorities, I'd say that it would be worth some extra effort to solve this problem without being dependent on speed limit data on road links.

I agree. For me, it's really awesome how far this app has come, and how much it provides while being GPL... but unfortunately these routing issues (this and the one that avoids traffic lights) make it so I can't use the app for my navigation. When it makes such strange routing decisions, I can't trust it to get me to my destination. The fact is that the map data is not perfect in the wild, and even when it is, heuristics on maxspeed aren't enough to prevent this kind of thing from happening in all cases. I'll have to keep using other navigation apps until this is fixed.

**Comment by vshcherb** [comment]
The cases has been reported to issue summary here https://github.com/osmandapp/Osmand/issues/9436. 
Please don't flood it with possible solutions and opinions and keep it clean to maintain known and already fixed use cases. 

If you want to help structure it, you are welcome.

**Comment by vshcherb** [comment]
This issue was cleaned from other regions to keep it clean for US

**Comment by gdt** [comment]
> @dead10ck This has been a longstanding problem. While the default maxspeed for a `motorway_link` is slightly lower than the default for a `motorway`, it's still higher than explicitly set maxspeed values in dense urban areas. A while back someone suggested making the OsmAnd default to be half of the default for motorway, but that never happened. (Personally I think that would be too low, anyway. Maybe 2/3 or 3/4 would be better.)
> 
> One thing you can do to help the problem is, if that exit ramp has a yellow speed sign, tag the _link with `maxspeed:advisory` and the value. OsmAnd will use that in place of a regulatory speed limit.

And, if the motorway has a white sign, put the maxspeed also on the link, because it applies.

I still think 1/2 is the right default.   It needs to be really certain that the assumed link speed is below any posted actual speed.  And, while a few links are very straight like this, many are not and are slower.   I routinely use a link that is between two motorways that are both 65 mph, with the link posted 20 mph.   It ought to be 35, but still that is basically half   Agreed that there are probably regional differences; in Mass there are highways shoehorned in around old cowpaths.

I also think that if there's an implied penalty (= routing dispreference for reasons other than expected time/distance) for links, that's good because there is more going on and it's irregular to use them for other than changing roads.

That said, I would agree to 2/3 as it's likely ok, and better than 1.

**Comment by vshcherb** [comment]
This issue can be fixed for United States if we know reasonable limit for motorway_link, so we can augment map data. But it is quite difficult to do cause in San Francisco default speed limit is 70 mph and this road is 65 mph.

**Comment by dead10ck** [comment]
In my experience, most freeways in the San Francisco Bay Area are 65mph, so that's probably a more reasonable default. Freeway exit ramps are more varied, but are usually around 35 mph. They are rarely greater than 45mph unless it's a link to another freeway.

Regardless of default maxspeeds, though, it should just never do this, even if there was some odd case where the motorway link had a faster speed limit. Relying on the freeway being faster isn't enough. If putting a penalty on exiting the freeway isn't feasible for some reason, then maybe this could be special-cased? If the route has you leave a freeway, then get right back on the same freeway, don't take that route.

**Comment by gdt** [comment]
If the link really is faster, and it's not painful to use it, it really is a better route.  Yes, that's pathological, but let's get an example before we solve a problem that doesn't exist.  The problem of default link speed assumptions is very real, and I think explains what we are seeing.

it is fair to have a penalty for being on a 1-line vs N-lane road, and for dealing with intersections and merges.  But still, I would like to see the main bug fixed, and then see what problems remain -- to me this is a core debugging principle.

**Comment by dead10ck** [comment]
I guess it depends on what you see the problem as. The way I see it, the problem has nothing to do with picking the fastest route. The app should never pick a route that gets off a road only to get right back on, even if it thought it was faster. It's unexpected and confusing, which is dangerous on the road because you have to divert your attention to try to figure out where in the world your navigator is taking you. This is the same reason #2625 is such a big problem too.

**Comment by vshcherb** [comment]
It also really depends what we want to achieve we have 2 open parts: map data and algorithms. So far it is not clear how is it possible to fix with algorithms, the only possible generic solution will be to depreciate all "*_links" which will cause navigation by motorways where links should be used (changing roads). 

Right now we are investigating what are the limits to depreciate links if that limit will be too high and it will cause another issue, there will be no other solution rather than introduce special tags for highways and we will get back to changing map data.

P.S. all routing algorithms are searching for the fastest route.

**Comment by gunnar22** [comment]
Maybe it's too costly CPU-wise, but all the needed information is already available: It should not navigate onto link ways unless it is routing away from the main highway for real and not coming right back to it. It should not make extremely sharp turns (close to 180 degrees) when there exists a more normalized alternative. These considerations are not focusing on fastest route, but on other rules. I think that something more than fastest route is needed in the algorithm. (However, maybe it will be too slow.)

It's true that all routing algorithms are searching for the fastest route, but are they doing ONLY that without any other rules?

**Comment by dead10ck** [comment]
This is not an unsolved problem. If it's not clear how it is possible to fix the algorithm, then perhaps studying other open source routing engines would help. There are several of them, and none of them have this problem. I think a good first step to solving this issue is acknowledging that it is, in fact, a problem, and that it is a problem unique to osmAnd's router.

> It's true that all routing algorithms are searching for the fastest route, but are they doing ONLY that without any other rules?

This is exactly my point. Saying "all routing algorithms are searching for the fastest route" is intellectually dishonest. @vshcherb you work on the router, and you are fully aware that there are lots of deciding factors in a router besides speed.

**Comment by gdt** [comment]
> I guess it depends on what you see the problem as. The way I see it, the problem has nothing to do with picking the fastest route. The app should never pick a route that gets off a road only to get right back on, even if it thought it was faster. It's unexpected and confusing, which is dangerous on the road because you have to divert your attention to try to figure out where in the world your navigator is taking you. This is the same reason #2625 is such a big problem too.
> 
> If we ask why the router does this, and the answer is "because it thought it was faster," then there's a big problem with the router. Speed is an important factor in choosing a route, but it is by no means the only factor, and maybe not even the most important.

I agree with you that getting off and on is bad.   I believe that routing based on min cost where cost is time is fundamentally sensible, but that in addition to estimating time for a route there should be a penalty applied for things that you don't want to do, even if they are faster.  By penalty I mean an extra time that does NOT represent the actual time taken, but the amount of time you'd rather drive to avoid doing something.    So if you find left turns scary, you might want to have a "30s penalty" for every left turn, so that a router that is 29s longer with one fewer left turn would be chosen.

So, I think penalties for getting off/on make sense, although they are hard to express locally in the calculations.

My main point was that I think almost all of this behavior is due to bad assumptions about link roads, and that if @vshcherb simply changes the default assumed speed for motorway_link to be half the default assumption for motorway (and so on), then I believe this "bad routing" problem will almost entirely disappear.  Only when that is done do I think it makes sense to assess how much residual off/on problems there are, and to understand what causes them and how to fix.

So far I have not seen a good reason not to change the motorway_link default speed assumption.  Yes, sometimes they are fast, but often they are curved and are not.  half speed (or 2/3) is a good guess in the presence of zero data.   full speed is almost always wrong.   As I said earlier, now that most links have speeds in MA, I just don't see this happen any more.

**Comment by dead10ck** [comment]
@gdt ah, thanks for clarifying; I think we're mostly on the same page. I think changing the default speeds and evaluating how much that solves the problem first is a reasonable approach. If it effectively solves the problem for all but the most rare of cases that hardly anyone will ever run into, then I agree it makes sense to prioritize something more general than heuristics just based on speed for later.

**Comment by vshcherb** [comment]
```
Saying "all routing algorithms are searching for the fastest route" is intellectually dishonest.
```
I don't think there is a technical background to assess my point how routing algorithm work. All routing algorithms are sorting routes by arriving time to destination point, sometimes this is different time that is displaying on the screen.

Lowering speed for all "_link" (including primary, tertiary, secondary) segments is definitely not correct cause it leads to wrong routing in cities where residential roads will be preferred over designated junctions. "motorway_link" / "trunk_link" could be another story.

This change is very easy to do in routing.xml (everybody can do it) but very difficult to test around world. Everybody who is volunteered to test it, you are welcome.

**Comment by scaidermern** [comment]
> Lowering speed for all "_link" (including primary, tertiary, secondary) segments is definitely not correct cause it leads to wrong routing in cities where residential roads will be preferred over designated junctions. "motorway_link" / "trunk_link" could be another story.

I agree that lowering the speed for all link roads is not a good idea. But what about lowering the speed only for motorway and trunk link roads?

**Comment by vshcherb** [comment]
```
			<select value="0.65" t="highway" v="motorway_link"/>
			<select value="0.65" t="highway" v="trunk_link"/>
```
Changes in the routing.xml for car navigation in priority section. 
So now we need to do thorough testing of this routing profile. 
https://github.com/osmandapp/OsmAnd-resources/blob/master/routing/routing_lowered_links.xml

Everybody can import this file and assign to car navigation and give it a try. If you do that, please share your feedback

**Comment by dead10ck** [comment]
Thanks @vshcherb. Sorry, but how exactly do I do that? When I went to `Settings | Import profile`, it complained that it needs to be an OSF extension, so I renamed the file to end in .osf. That got it to get past the file picker, but then nothing happens. There's no new profile, and I can't tell if anything has been done. I tried exporting my existing profile, and I tried to look at it, but it appears to be a zip file with two JSON files in it, which tells me I may be doing the complete wrong thing.

**Comment by gunnar22** [comment]
Here is one way: Shut down OsmAnd, copy the file to a PC, edit the file in a text editor, copy it back (to replace the original on the phone), start OsmAnd.

**Comment by jfburke3** [comment]
@dead10ck @gunnar22  I think that @vshcherb has already made the changes in that file. Just copy it to your phone, and put it in the OsmAnd data directory, then rename it to routing.xml.  Close OsmAnd (maybe even force stop it to be sure). When you restart it, it will detect the file and process it. 

That said, having spent a lot of time thinking about the problem, I'm kinda with @vshcherb on it, although having the higher default speed on motorway links can have issues a lot of the time. 

I've encountered the problem a lot, and one thing I've done that has helped is to edit OSM and put traffic signals, stop signs, or yield signs, as appropriate, at the end of the exits.

**Comment by dead10ck** [comment]
Thanks @jfburke3. I tried putting the file in the data directory that is displayed in the storage location of the settings, and force closing the app. After doing so, the problem was still there navigating between the two points listed in this issue.

Is there a way I can confirm that the routing.xml is being detected and processed? I couldn't find any other existing routing.xml file in the data directory.

> I've encountered the problem a lot, and one thing I've done that has helped is to edit OSM and put traffic signals, stop signs, or yield signs, as appropriate, at the end of the exits.

Yes, indeed, if the map data were perfectly complete and accurate, we wouldn't have as many problems. The problem is: if I run into one of these problems that gives the router an issue, as someone who has the knowledge and willingness to go correct the map data at the particular location that caused the problem, I can certainly fix the data so that when I take that route again, I get a sane route. But that does not help me when I go somewhere new for the first time that encounters one of these situations, and it doesn't help me if I never take that route again. And I would be willing to bet that most of your users are not people with the knowledge and willingness to go fix the map data. Most people just want a navigation app that works and don't care how or why. If it takes them off the freeway and gets back on, they don't care if it's because the maxspeed tag was missing, and the default speed was way too high. They've already had a bad experience, and it makes them less inclined to use it again.

The fact is that the map data is not complete and perfect. If you want the navigation routes to be useful and safe, you need to work with the data you have and perform heuristics.

**Comment by vshcherb** [comment]
Sorry was away. You don't need to overwrite routing.xml!
You can put it in the folder (import routing file)
![image](https://user-images.githubusercontent.com/1042025/90317265-58386c00-df28-11ea-8cc2-b709212ea4e8.png)

**Comment by dead10ck** [comment]
@vshcherb how do I get to this screen? I've looked through the menus and can't find that. The closest thing I can find is the "Navigation settings" page, and this is not there.

![Screenshot_20200815-143353](https://user-images.githubusercontent.com/846275/90319275-96bc2f80-df04-11ea-9065-b414a47d0c93.jpg)

**Comment by vshcherb** [comment]
@dead10ck Navigation Settigns  -> Navigation Type -> scroll down...

**Comment by dead10ck** [comment]
@vshcherb ah, thanks, I didn't realize the "Navigation Settings" page was only reachable through the Navigation button, and then clicking "Options." Plus, there is another page buried in the main Settings menu called "Navigation Settings" which has entirely different options, which is a little confusing.

Anyway, I imported the routing.xml and the problem goes away between the two points listed in the first post in this issue! 🎉

**Comment by vshcherb** [comment]
@dead10ck the idea to check if there is no other issues appeared :-) I'm thinking how to check now cause we don't have much time before release though change is pretty simple and probably we will put it in without months of testing.

**Comment by ErichRitz** [comment]
I wanted to chime in and say the same issue appears in 2 places along I-17 northbound headed north out of Phoenix, AZ towards Flagstaff:
https://www.openstreetmap.org/directions?engine=graphhopper_car&route=34.0205%2C-112.1436%3B34.1442%2C-112.1412#map=16/34.0839/-112.1409&layers=C

![Screenshot_2020-08-18-17-01-02](https://user-images.githubusercontent.com/5106082/90668881-945c1d00-e205-11ea-9796-964985ed09f7.png)

![Screenshot_2020-08-18-17-01-35](https://user-images.githubusercontent.com/5106082/90668901-9a51fe00-e205-11ea-8e77-ca0e65280a76.png)

You can see that grasshopper routes correctly (follow the openstreetmap.org link above).

Using the provided routing_lowered_links.xml solves the problem.  Additionally, routing from my house to the interstate remains the same (i.e., there are no adverse affects for routing from surface streets onto the interstate).  But, I have not extensively tested the new routing_lowered_links.xml.

In this particular instance the speed limit is set to 65 mph for the interstate, the *_links do not have maxspeed set, and the *_link intersections do not have stop signs in OSM (although stop signs are visible in Google Map Street View, I have not personally taken those exits so technically I'm not allowed to add them to OSM).

I'll also echo the comments that in my experience, in the US, _links* rarely have maxspeed set.
